src/transactions/PathPaymentOpFrame.cpp

```diff
@@ -282,24 +280,31 @@ PathPaymentOpFrame::doApply(Application& app, LedgerDelta& delta,
             return false;
         }
 
-        auto ok = sourceAccount->addBalance(-curBSent);
+        if (getCurrentLedgerVersion(header) < 8)
+        {
+            sourceAccount.forget(ls);
+            auto thisAccount = stellar::loadAccountRaw(ls, getSourceID());
+            if (!thisAccount)
+            {
+                throw std::runtime_error("modifying account that does not exist");
+            }
+            thisAccount->invalidate();
+            sourceAccount = loadSourceAccount(ls, header);
+        }
+        header->invalidate();
+
+        auto ok = sourceAccount.addBalance(-curBSent);
         assert(ok);
-        sourceAccount->storeChange(delta, db);
     }
     else
     {
-        TrustFrame::pointer sourceLineFrame;
-        if (bypassIssuerCheck)
+        if (!bypassIssuerCheck)
         {
-            sourceLineFrame =
-                TrustFrame::loadTrustLine(getSourceID(), curB, db, &delta);
-        }
-        else
-        {
-            auto tlI =
-                TrustFrame::loadTrustLineIssuer(getSourceID(), curB, db, delta);
-
-            if (!tlI.second)
+            // We need a LedgerState here since it is possible that changes to
+            // issuer were already stored.
```

Comment:
comment is misleading, I think what you mean is that issuer may be one of the other accounts used in this function (such as `sourceAccount`), in which case, isn't it easier to just invalidate after changing the balance of the `sourceAccount`? (RAII to the rescue)